src/vec.rs

```diff
@@ -0,0 +1,364 @@
+    }
+
+    pub fn clear(&mut self) {
+        self.len = 0;
+        self.save();
+    }
+
+    #[inline]
+    fn index_to_entry_addr(&self, index: usize) -> Address {
+        self.base + Bytes::from((Self::ENTRY_SIZE * index) as u64)
+    }
+
+    #[inline]
+    fn read_value(&self, index: usize) -> T {
+        let mut p = self.index_to_entry_addr(index);
+        // ret = read_struct(ptr, &self.memory);
+        let mut buf = if T::FIXED_SIZE {
+            vec![0; T::MAX_SIZE as usize]
+        } else {
+            let size = read_u32(&self.memory, p) as usize;
+            p += Bytes::from(4u64);
+            vec![0; size]
+        };
+        self.memory.read(p.get(), &mut buf);
+        T::from_bytes(buf)
+    }
+
+    #[inline]
+    fn write_value(&mut self, index: usize, value: &T) {
+        let value = value.to_bytes();
+        let mut p = self.index_to_entry_addr(index);
+        if !T::FIXED_SIZE {
+            write_u32(&self.memory, p, value.len() as u32);
+            p += Bytes::from(4u64);
+        }
+        self.memory.write(p.get(), &value);
+    }
+
+    pub fn insert(&mut self, index: usize, element: &T) {
+        let len = self.len();
+        assert!(index <= len);
+
+        // space for the new element
+        if len == self.capacity() {
+            self.reserve(1);
+        }
+
+        // Move all entries starting at index one to the right.
+        let src = self.index_to_entry_addr(index);
+        let dst = self.index_to_entry_addr(index + 1);
+        copy(
+            &self.memory,
+            src,
+            dst,
+            (len - index) as u64,
+            Self::ENTRY_SIZE,
+        );
+
+        self.write_value(index, element);
+
+        self.len += 1;
+        self.save();
+    }
+
+    pub fn remove(&mut self, index: usize) -> T {
+        let len = self.len();
+        assert!(index < len);
+
+        let ret = self.read_value(index);
+
+        // Shift everything down to fill in that spot.
+        copy::<M>(
+            &self.memory,
+            self.index_to_entry_addr(index + 1),
+            self.index_to_entry_addr(index),
+            (len - index - 1) as u64,
+            Self::ENTRY_SIZE,
+        );
+
+        self.len -= 1;
+        self.save();
+        ret
+    }
+
+    #[inline]
+    pub fn push(&mut self, value: &T) {
+        self.write_value(self.len(), value);
+        self.len += 1;
+        self.save();
+    }
+
+    #[inline]
+    pub fn pop(&mut self) -> Option<T> {
+        if self.len == 0 {
+            None
+        } else {
+            self.len -= 1;
+            self.save();
+            Some(self.read_value(self.len()))
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
+    /// even if the resulting reference is not used.
+    pub unsafe fn get_unchecked(&self, index: usize) -> T {
+        self.read_value(index)
+    }
```

Comment:
What is the use-case for this endpoint?

Reply:
If I know that the access in bounds. For example in the binary heap, I do the exact same check.